src/derivative_utils.jl

```diff
@@ -54,10 +54,161 @@ function calc_J!(integrator, cache::OrdinaryDiffEqMutableCache, is_compos)
+
+    WOperator(f,gamma[;transform=false])
+
+`f` needs to have a jacobian and `jac_prototype`, but the prototype does not need
+to be a diffeq operator --- it will automatically be converted to one.
+
+`WOperator` supports lazy `*` and `mul!` operations, the latter utilizing an
+internal cache (can be specified in the constructor; default to regular `Vector`).
+It supports all of `AbstractDiffEqLinearOperator`'s interface.
+"""
+mutable struct WOperator{T,
+  MType <: Union{UniformScaling,AbstractMatrix},
+  GType <: Real,
+  JType <: DiffEqBase.AbstractDiffEqLinearOperator{T}
+  } <: DiffEqBase.AbstractDiffEqLinearOperator{T}
+  mass_matrix::MType
+  gamma::GType
+  J::JType
+  transform::Bool       # true => W = mm/gamma - J; false => W = mm - gamma*J
+  _func_cache           # cache used in `mul!`
+  _concrete_form         # non-lazy form (matrix/number) of the operator
+  WOperator(mass_matrix, gamma, J; transform=false) = new{eltype(J),typeof(mass_matrix),
+    typeof(gamma),typeof(J)}(mass_matrix,gamma,J,transform,nothing,nothing)
+end
+function WOperator(f::DiffEqBase.AbstractODEFunction, gamma; transform=false)
+  @assert DiffEqBase.has_jac(f) "f needs to have an associated jacobian"
+  if isa(f, Union{SplitFunction, DynamicalODEFunction})
+    error("WOperator does not support $(typeof(f)) yet")
+  end
+  # Convert mass matrix, if needed
+  mass_matrix = f.mass_matrix
+  if !isa(mass_matrix, Union{AbstractMatrix,UniformScaling})
+    mass_matrix = convert(AbstractMatrix, mass_matrix)
+  end
+  # Convert jacobian, if needed
+  J = deepcopy(f.jac_prototype)
+  if !isa(J, DiffEqBase.AbstractDiffEqLinearOperator)
+    J = DiffEqArrayOperator(J; update_func=f.jac)
+  end
+  return WOperator(mass_matrix, gamma, J; transform=transform)
+end
+
+set_gamma!(W::WOperator, gamma) = (W.gamma = gamma; W)
+DiffEqBase.update_coefficients!(W::WOperator,u,p,t) = (update_coefficients!(W.J,u,p,t); W)
+function Base.convert(::Type{AbstractMatrix}, W::WOperator)
+  if W._concrete_form == nothing
+    # Allocating
+    if W.transform
+      W._concrete_form = W.mass_matrix / W.gamma - convert(AbstractMatrix,W.J)
+    else
+      W._concrete_form = W.mass_matrix - W.gamma * convert(AbstractMatrix,W.J)
+    end
+  else
+    # Non-allocating
+    if W.transform
+      rmul!(copyto!(W._concrete_form, W.mass_matrix), 1/W.gamma)
+      axpy!(-1, convert(AbstractMatrix,W.J), W._concrete_form)
+    else
+      copyto!(W._concrete_form, W.mass_matrix)
+      axpy!(-W.gamma, convert(AbstractMatrix,W.J), W._concrete_form)
+    end
+  end
+  W._concrete_form
+end
+function Base.convert(::Type{Number}, W::WOperator)
+  if W.transform
+    W._concrete_form = W.mass_matrix / W.gamma - convert(Number,W.J)
+  else
+    W._concrete_form = W.mass_matrix - W.gamma * convert(Number,W.J)
+  end
+  W._concrete_form
+end
+Base.size(W::WOperator, args...) = size(W.J, args...)
+function Base.getindex(W::WOperator, i::Int)
+  if W.transform
+    W.mass_matrix[i] / W.gamma - W.J[i]
```

Comment:
Since this shows up everywhere and it's constant given by the algorithm, it might make sense to make this a type parameter so the code can simply dispatch off of it, or at least not have large performance cuts because of this conditioning in indexing. I'm not sure this indexing is really used all that much though.

Reply:
> I'm not sure this indexing is really used all that much though.

This is true both for factorization (it indexes the concrete matrix instead of W) and `mul!` (which is lazy).

Reply:
Sounds good